### ISSUE TITLE
Proposal: Component Keys

### ISSUE THREAD

**Issue by erickzanardo** [flame-engine/flame#2443]
# What could be improved
Finding a specific component in the component tree.

# Why should this be improved
Right now, we have a couple of ways of finding components in the component tree in a `FlameGame`, like the `firstChild`, `descendants().where(...)`, `where` and so on.

But, if we need to find a specific component in the tree, and do not worry about possible perfomance issues like iterating of the descendants of the game of a component node, developers usually have to keep a reference of it manually somewhere in their code.

Although this is not complicated, I feel this could be improved.

# Any risks?
None that I can think of.

# More information
My proposal is to introduce a similar API of the Flutter keys on widgets. In this proposal, components could receive an optional `Key` instance (or probably `ComponentKey` so we don't clash with Flutter's class), and when that is provided, that component could be retrieved directly from the game instance in an API like this:

```
final player = gameRef.getComponentById(const ComponentKey('player'));
```
We could also make that method generic for better type safeness

```
final player = gameRef.getComponentById<Player>(const ComponentKey('player'));
```

I haven't thought much on the implementation, but my (possible naive) suggestion is that we could simply keep a map in the game, and on the `Component.onMount` method, if a key was supplied, that component would put a reference of itself on the map using that key, and on its `onRemove`, it would be removed from that map.

That way, getting a component by key would always be a very cheap operation and I also believe that this API may have the bonus of being similar to Flutter's one, which may help Flutter developers that aren't familiar with Flame yet.

Looking forward for feedbacks on this.

**Comment by luanpotter** [comment]
Why not just keep a reference to the component itself, instead of keeping reference to a key, that then can be mapped to the component?

**Comment by ufrshubham** [comment]
> Why not just keep a reference to the component itself, instead of keeping reference to a key, that then can be mapped to the component?

Wouldn't that stop the GC from freeing removed and unreferenced components?

**Comment by luanpotter** [comment]
It would, but if you are still holding a key to it, don't you still want to keep access to it? Maybe I am just not seeing the user case, would maybe components be created under the same key at different times?

**Comment by ufrshubham** [comment]
Okay, I am imagining it to be something similar to Unity's [GameObject.FindWithTag](https://docs.unity3d.com/ScriptReference/GameObject.FindWithTag.html) or roughly similar to Godot's [get_node](https://docs.godotengine.org/en/stable/classes/class_node.html#class-node-method-get-node). 

I'd expect it to return a component at a given key if it is present in the component tree, null otherwise. Basically users shouldn't worry about when it is added/removed from the map (ideally they shouldn't even care how we are storing and finding the components internally).

**Comment by st-pasha** [comment]
Can the key be a simple string, like `"player"`?
The global cache can utilize weak references (though if we're going to remove a component from cache in `onRemove`, that's hardly needed).

**Comment by erickzanardo** [comment]
> Okay, I am imagining it to be something similar to Unity's [GameObject.FindWithTag](https://docs.unity3d.com/ScriptReference/GameObject.FindWithTag.html) or roughly similar to Godot's [get_node](https://docs.godotengine.org/en/stable/classes/class_node.html#class-node-method-get-node). 
> 
> 
> 
> I'd expect it to return a component at a given key if it is present in the component tree, null otherwise. Basically users shouldn't worry about when it is added/removed from the map (ideally they shouldn't even care how we are storing and finding the components internally).

Exactly what I was thinking!

**Comment by erickzanardo** [comment]
> Can the key be a simple string, like `"player"`?
> 
> The global cache can utilize weak references (though if we're going to remove a component from cache in `onRemove`, that's hardly needed).
> 
> 

We can, I am ok with tbh. Thought on the class thing so we can mimic flutter api and maybe make it more comfortable for flutter devs

**Comment by luanpotter** [comment]
I think it could be really useful as well if you are able to access it back _without_ holding a reference to the key. I.e. either a string or a "value class" that you can re-create with the same string later.

That being said, for the cases that you do hold a ref to the key, I have a small suggestion for improvement: put the generic type (optionally) in the key. So similar to how HasGameRef work/used to work. You could have a:

```
final key = GenericKey("foo");
final player = GenericKey<Player>("player");
```

so in the second case you don't need to specify the type when getting it. not a big difference either way, just a thought.

**Comment by erickzanardo** [comment]
Implemented in #2566
